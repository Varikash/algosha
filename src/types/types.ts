import { ElementStates } from "./element-states";

export type StringElement = {
  value: string;
  state: ElementStates;
}