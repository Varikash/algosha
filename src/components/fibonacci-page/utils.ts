import { delayExecution } from "../../constants/utils";
import { DELAY_IN_MS } from "../../constants/delays";


export async function generateFibonacciSequence(
  n: number,
  setSequence: React.Dispatch<React.SetStateAction<number[]>>,
  setLoading: React.Dispatch<React.SetStateAction<boolean>>,
  setInputValue: React.Dispatch<React.SetStateAction<number | undefined>>,
  ): Promise<number[]> {
  if (n <= 0) {
    return [];
  }
  let sequence = [1];
  setSequence([...sequence])
  if (n > 1) {
    await delayExecution(DELAY_IN_MS)
    sequence.push(1);
    setSequence([...sequence])
  }
  for (let i = 2; i <= n; i++) {
    await delayExecution(DELAY_IN_MS)
    sequence.push(sequence[i - 1] + sequence[i - 2]);
    setSequence([...sequence])
  }
  setLoading(false);
  setInputValue(undefined);
  return sequence;
}