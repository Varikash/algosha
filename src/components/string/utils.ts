import { StringElement } from "../../types/types";
import { ElementStates } from "../../types/element-states";
import { delayExecution } from "../../constants/utils";
import { DELAY_IN_MS } from "../../constants/delays";
import React from "react";

const swap = (arr: StringElement[], firstIndex: number, secondIndex: number) => {
  [arr[firstIndex], arr[secondIndex]] = [arr[secondIndex], arr[firstIndex]];
};


export const swapStrings = async (
  arr: StringElement[],
  setArray: React.Dispatch<React.SetStateAction<StringElement[]>>,
  setLoading: React.Dispatch<React.SetStateAction<boolean>>,
  setState: React.Dispatch<React.SetStateAction<boolean>>
  ) => {

  let length = arr.length - 1;
  let half = Math.floor(length / 2);

  for (let head = 0, tail = length; head < tail; head++, tail--) {

    arr[head].state = ElementStates.Changing;
    arr[tail].state = ElementStates.Changing;
    setArray([...arr]);

    await delayExecution(DELAY_IN_MS);


    swap(arr, head, tail);


    arr[head].state = ElementStates.Modified;
    arr[tail].state = ElementStates.Modified;
    setArray([...arr]);

    await delayExecution(DELAY_IN_MS);
  }

  if (arr.length % 2 !== 0) {
    arr[half].state = ElementStates.Modified;
    setArray([...arr]);
  }

  setLoading(false);
  setState(true);
};