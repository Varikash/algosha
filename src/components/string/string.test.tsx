import { swapStrings } from './utils';
import { StringElement } from "../../types/types";
import { ElementStates } from "../../types/element-states";

describe("reverse string", () => {

  jest.setTimeout(10000); // Без тайм-аута не успевает сработать(

  const mockSetArray = jest.fn();
  const mockSetLoading = jest.fn();
  const mockSetState = jest.fn();

  test.each([
    ["even string length", "winter", "retniw"],
    ["odd string length", "спасите", "етисапс"],
    ["one char", "ы", "ы"],
    ["empty string", "", ""],
  ])("%s", async (_, initialValue: string, expectedValue: string) => {
    const stringArray: StringElement[] = initialValue.split('').map((char) => ({
      value: char,
      state: ElementStates.Default
    }));

    await swapStrings(stringArray, mockSetArray, mockSetLoading, mockSetState);

    if (mockSetArray.mock.calls.length > 0) {
      const finalCall = mockSetArray.mock.calls[mockSetArray.mock.calls.length - 1][0];
      const finalArray = finalCall as StringElement[];
      const resultString = finalArray.map((element: StringElement) => element.value).join('');

      expect(resultString).toBe(expectedValue);
      expect(finalArray).toMatchSnapshot();
    } else {
      expect(expectedValue).toBe('');
      expect(mockSetArray).not.toHaveBeenCalled();
    }
  });
});