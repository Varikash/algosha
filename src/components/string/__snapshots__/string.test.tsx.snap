// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reverse string even string length 1`] = `
Array [
  Object {
    "state": "modified",
    "value": "r",
  },
  Object {
    "state": "modified",
    "value": "e",
  },
  Object {
    "state": "modified",
    "value": "t",
  },
  Object {
    "state": "modified",
    "value": "n",
  },
  Object {
    "state": "modified",
    "value": "i",
  },
  Object {
    "state": "modified",
    "value": "w",
  },
]
`;

exports[`reverse string odd string length 1`] = `
Array [
  Object {
    "state": "modified",
    "value": "е",
  },
  Object {
    "state": "modified",
    "value": "т",
  },
  Object {
    "state": "modified",
    "value": "и",
  },
  Object {
    "state": "modified",
    "value": "с",
  },
  Object {
    "state": "modified",
    "value": "а",
  },
  Object {
    "state": "modified",
    "value": "п",
  },
  Object {
    "state": "modified",
    "value": "с",
  },
]
`;

exports[`reverse string one char 1`] = `
Array [
  Object {
    "state": "modified",
    "value": "ы",
  },
]
`;
